### PR TITLE
Use straight deltas for the exact conversion.

### PR DESCRIPTION
**Pull request by Arthur2e5**
See #24 comments.

> As a side note, commit 86ae326 shows that applying straight deltas converges faster than the current bisection-like behavior. This might be true since the approximate scale between the two coord systems are still (almost perfectly) preserved.
>
> I will open a separate PR for that optimization, and you can pick the better one. I will try implementing the one you selected for the rest of the languages too.

### CODE REVIEW

**Comment by Arthur2e5** [comment]
Looks like these are essentially the same…Let me use delta.

**Comment by Arthur2e5** [comment]
Assuming this is the better one, beginning to do similar things to other languages…

**Comment by Arthur2e5** [comment]
Not touching C until #27 gets merged.

File: javascript/transform.js [hunk 1]
```diff
@@ -15,27 +16,30 @@ function outOfChina(lat, lng) {
...
+	ret += 20.0*Math.sin(y*Math.PI) + 40.0*Math.sin(y/3.0*Math.PI);
+	ret += 160.0*Math.sin(y/12.0*Math.PI) + 320*Math.sin(y*Math.PI/30.0);
+	ret *= 2.0 / 3.0;
+	ret += -100.0 + 2.0*x + 3.0*y + 0.2*y*y + 0.1*x*y + 0.2*Math.sqrt(Math.abs(x));
 	return ret;
 }
 
-function transformLon(x, y) {
-	var ret = 300.0 + x + 2.0*y + 0.1*x*x + 0.1*x*y + 0.1*Math.sqrt(Math.abs(x));
-	ret += (20.0*Math.sin(6.0*x*Math.PI) + 20.0*Math.sin(2.0*x*Math.PI)) * 2.0 / 3.0;
```
Comment (Arthur2e5) [inline javascript/transform.js]: The most terrible part of writing the long expr this way is that compilers will not optimize this section to respect IEEE754…

**Comment by Arthur2e5** [comment]
Well, I will leave many of these impls until I get confirmed that nobody is using the mis-exported internal functions…

File: javascript/transform.js [hunk 2]
```diff
@@ -15,27 +16,30 @@ function outOfChina(lat, lng) {
 	return false;
 }
 
-function transformLat(x, y) {
-	var ret = -100.0 + 2.0*x + 3.0*y + 0.2*y*y + 0.1*x*y + 0.2*Math.sqrt(Math.abs(x));
-	ret += (20.0*Math.sin(6.0*x*Math.PI) + 20.0*Math.sin(2.0*x*Math.PI)) * 2.0 / 3.0;
-	ret += (20.0*Math.sin(y*Math.PI) + 40.0*Math.sin(y/3.0*Math.PI)) * 2.0 / 3.0;
-	ret += (160.0*Math.sin(y/12.0*Math.PI) + 320*Math.sin(y*Math.PI/30.0)) * 2.0 / 3.0;
+function gcjTrigLat(x, y) {
```
Comment (Arthur2e5) [inline javascript/transform.js]: Hey why do people ever try to make this a separate function…

File: javascript/transform.js [hunk 3]
```diff
@@ -58,36 +62,24 @@ function gcj2wgs(gcjLat, gcjLng) {
...
-		if (dLat > 0) {
-			pLat = wgsLat;
-		} else {
-			mLat = wgsLat;
-		}
-		if (dLng > 0) {
-			pLng = wgsLng;
-		} else {
-			mLng = wgsLng;
+		var gcjDiff = gcjDelta(wgsLat, wgsLng);
```
Comment (googollee) [inline javascript/transform.js]: Seems you change algorithm. Could you describe it?

Comment (Arthur2e5) [inline javascript/transform.js]: Already explained in the comments and PR desc. Also read the R description.

Comment (Arthur2e5) [inline javascript/transform.js]: https://github.com/caijun/geoChina/blob/master/R/cst.R#L124

Comment (Arthur2e5) [inline javascript/transform.js]: And, good morning meow.

**Comment by Arthur2e5** [comment]
Personally I would prefer to make many operands `struct`-or-`class`-or-`Object`-or-`tuple`-ized, but that seems to introduce incompatibility on commonly-used funcs.

File: javascript/transform.js (hunk 3 above)
Comment (googollee) [inline javascript/transform.js]: I will try to understand the R description during sprint festival. Thanks. If it's OK, I will try to merge it to other language.

Woof.

Comment (Arthur2e5) [inline javascript/transform.js]: A much more aggressive version is already in the C version.

File: javascript/transform.js (hunk 2 above)
Comment (googollee) [inline javascript/transform.js]: It improves readability of code.

Comment (Arthur2e5) [inline javascript/transform.js]: Yeah, size-based readers like V8 would be more happy to optimize small functions.

Humans have infinitely tall screens anyway.

**Pull request merged**